apply_DDT.py

```diff
@@ -0,0 +1,594 @@
+            ax.hist(mT[score_mask], bins=50, range=(180,650), histtype='step', label=f'DDT(BDT cut = {cuts})')
+ 
+        #ax.set_title('mT for Different DDT(BDT) Thresholds')
+        ax.set_xlabel('$\\mathrm{m}_{\\mathrm{T}}$')
+        ax.set_ylabel('Events')
+        ax.legend()
+        plt.tight_layout()
+        plt.savefig('plots/bkg_events_vs_mT.png')
+        plt.savefig('plots/bkg_events_vs_mT.pdf')
+        
+        # log scale it
+        ax.set_yscale('log')
+        plt.tight_layout()
+        plt.savefig('plots/log_bkg_events_vs_mT.png')
+        plt.savefig('plots/log_bkg_events_vs_mT.pdf')
+ 
+        # Do it again normalized to unit area
+        fig, ax = plt.subplots(figsize=(10, 8))
+        # Options to make the plot fancier 
+        hep.cms.label(rlabel="2018 (13 TeV)")
+        for cuts, scores in zip(bdt_cuts, BKG_score_ddt):
+            score_mask = scores > 0.0 # DDT score > 0.0 is equivalent to BDT score about BDT cut value
+            ax.hist(mT[score_mask], bins=50, range=(180,650), histtype='step', label=f'DDT(BDT cut = {cuts})', density=True)
+ 
+        #ax.set_title('mT for Different DDT(BDT) Thresholds')
+        ax.set_xlabel('$\\mathrm{m}_{\\mathrm{T}}$')
+        ax.set_ylabel('Events')
+        ax.legend()
+        plt.tight_layout()
+        plt.savefig('plots/norm_bkg_events_vs_mT.png')
+        plt.savefig('plots/norm_bkg_events_vs_mT.pdf')
+        plt.close()
+ 
+        print(BKG_score_ddt) 
+
+    # _____________________________________________
+    # Create Significance Plots
+    if plot_fom_significance == True :
+
+        # Grab the signal data
+        sig_cols = [Columns.load(f) for f in glob.glob(sig_data_files)]
+ 
+        # Prepare a figure
+        fig = plt.figure(figsize=(10, 7))
+        hep.cms.label(rlabel="2018 (13 TeV)")
+        ax = fig.gca()
+
+        best_bdt_cuts = [] #store the best bdt values
+        # Iterate over the variables in the 'con' dictionary
+        for sig_col in sig_cols:
+            mz = sig_col.metadata['mz']
+            #if mz not in [300, 350, 450, 550]:
+            #  continue
+            s = f'bsvj_{mz:d}_10_0.3'
+         
+            # Grab the input features and weights
+            sig_X = sig_col.to_numpy(all_features)
+            sig_weight = sig_col.xs / sig_col.cutflow['raw'] * lumi * sig_col.arrays['puweight']
+      
+            # Apply the signal region
+            sig_rt = sig_X[:,-1]
+            sig_X = sig_X[sig_rt > 1.18]
+            sig_weight = sig_weight[sig_rt > 1.18]
+            print("M(Z') = ", mz, " Events: ", len(sig_X), " weights: ", sig_weight)
+  
+            # grab rt (after cut)
+            sig_rt = sig_rt[sig_rt>1.18]
+            sig_X = sig_X[:,:-1]
+      
+            # grab rho
+            sig_rho = sig_X[:,-1]
+            sig_X = sig_X[:,:-1] # remove it from X so that eventually it can be used for BDT scores
```

Comment:
it seems like there's a fair amount of code duplication (e.g. these lines, but also others). can some of these duplicated procedures become functions that are called whenever needed?

Reply:
Yes... I realized that when I made slight changes each time... I wish I had thought to do that initially, I will go ahead and do that now.